pkg/reference/ref.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reference
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+var (
+	// DefaultParentKind is the default metav1.GroupKind for ParentReferences.
+	DefaultParentKind = metav1.GroupKind{
+		Group: v1alpha2.GroupName,
+		Kind:  "Gateway",
+	}
+
+	// DefaultSecretKind is the default metav1.GroupKind for SecretObjectReferences.
+	DefaultSecretKind = metav1.GroupKind{
+		Group: "",
+		Kind:  "Secret",
+	}
+
+	// DefaultBackendObjectKind is the default metav1.GroupKind for BackendObjectReferences.
+	DefaultBackendObjectKind = metav1.GroupKind{
+		Group: "",
+		Kind:  "Service",
+	}
+)
+
+func referenceKind(gk metav1.GroupKind, g *v1alpha2.Group, k *v1alpha2.Kind) metav1.GroupKind {
+	if g != nil {
+		gk.Group = string(*g)
+	}
+
+	if k != nil {
+		gk.Kind = string(*k)
+	}
+
+	return gk
+}
+
+func referenceNamespace(ns *v1alpha2.Namespace) string {
+	if ns != nil {
+		return string(*ns)
+	}
+
+	return ""
+}
+
+// Ref is an interface that represents a reference to a resource.
+type Ref interface {
+	Name() string
+	Namespace() string
+	Kind() metav1.GroupKind
+}
+
+type localRef struct {
+	ref *v1alpha2.LocalObjectReference
+}
+
+func (l *localRef) Name() string {
+	return string(l.ref.Name)
+}
+
+func (l *localRef) Namespace() string {
+	return ""
+}
+
+func (l *localRef) Kind() metav1.GroupKind {
+	return metav1.GroupKind{Group: string(l.ref.Group), Kind: string(l.ref.Kind)}
+}
+
+// Local returns a Ref for the LocalObjectReference ref.
```

Comment:
Nit: When reading through the code, I kept on thinking that this indicated the start of a section, when it actually indicates the end. It might be slightly easier to read/understand if these functions preceded everything else about the type. Feel free to disregard though.

Reply:
Fixed.